datalad_next/constraints/base.py

```diff
@@ -0,0 +1,164 @@
+""""""
+__docformat__ = 'restructuredtext'
+
+from typing import (
+    TYPE_CHECKING,
+    TypeVar,
+)
+
+if TYPE_CHECKING:  # pragma: no cover
+    from datalad.distribution.dataset import Dataset
+
+aConstraint = TypeVar('aConstraint', bound='Constraint')
+aDataset = TypeVar('aDataset', bound='Dataset')
```

Comment:
Just a remark: I think the convention for type var naming is capitalized. Also: in  OO-languages "aX"  usually refers to an instance of "X". Although this is not PEP-8 style ---which would be "a_x"--- the notation might be confusing to some people, I don't have an obviously superior recommendation though. "DatasetType", "DatasetLike", ...?

Also, I am not sure whether the type variables are necessary here. Type-checking should work with subclasses.

Reply:
Can you clarify please? `aDataset` is meant to refer to an instance.

The purpose of type vars here is to avoid the import of the underlying class. To avoid circularity (in case of `Constraint`) or to avoid making the constraint system depending on "all of datalad" (in case of `Dataset`).

Reply:
FTR: here is the PEP8 snippet https://peps.python.org/pep-0008/#type-variable-names

Reply:
I guess `DatasetDerived` would somehow reflect the semantics of `bound=`